bench.c

```diff
@@ -0,0 +1,189 @@
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#define TARGET_HOST "127.0.0.1"
+#define TARGET_PORT 12345
+#define BENCH_COUNT 10
+#define RESULT_FILE_NAME "bench.txt"
+
+/* vary-length message should shorter than this */
+#define MAX_MSG_LEN 32
+
+/*
+ * Too much concurrent connection would be treat as sort of DDOS attack
+ * (mainly caused by configs (kernel: `tcp_max_syn_backlog` and
+ * `somaxconn`. Application (fastecho): `backlog`). Nevertheless, default
+ * maximum number of fd per-process is `1024`, if you
+ * insist to proceed the benchmark, perform following modifications:
+ *       (PROCEDURE MAY VARY BETWEEN DIFFERENT DISTROS)
+ * For Ubuntu:
+ * (1)
+ * Use following commands to adjust kernel attributes:
+ *     a. `$ sudo sysctl net.core.somaxconn={depends_on_MAX_THREAD}`
+ *     b. `$ sudo sysctl net.ipv4.tcp_max_syn_backlog={ditto}`
+ * Note that `$ sudo sysctl net.core.somaxconn` can get current value.
+ * `somaxconn` is max amount of established connection, whereas
+ * `tcp_max_syn_backlog` is max amount of connection at first step
+ * of TCP 3-way handshake (SYN).
+ *
+ * (2)
+ * Use command `$ ulimit -n {ditto}` to enlarge limitation of fd
+ * per-process. Note that this modification only effects on process
+ * which executes the command and its child processes.
+ *
+ * (3)
+ * Specify `backlog` with value as large as `net.ipv4.tcp_max_syn_backlog`.
+ *
+ * Remember to reset the modifications after benchmarking to keep
+ * stability of running machine
+ */
+#define MAX_THREAD 1000
+
+/*
+ * TODO: use vary-length msg instead would be better since it's more
+ * conforms to reality
+ */
+const char *msg_dum = "dummy message";
+
+pthread_t pt[MAX_THREAD];
+
+/* block all worker before they are all ready to bombing kecho */
+bool ready;
+
+pthread_mutex_t res_lock = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
`res_lock` should be declared as a local variable rather than a global one.

Reply:
I just tried this, and it works correctly. But I'm confused, doesn't a local lock only used by worker itself instead of shared by all workers?

Reply:
Worker threads should not preserve much resources at first glance. In this case, it does not matter at all.